include/boost/histogram/display.hpp

```diff
@@ -0,0 +1,212 @@
+#ifndef DISPLAY_HPP
+#define DISPLAY_HPP
+
+#include <boost/format.hpp>
+#include <boost/histogram.hpp>
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace display {
+namespace {
+using namespace boost::histogram;
+
+const unsigned int histogram_width = 60; // 60 characters
+const float max_bin_coefficient = 0.95;  // 95% of histogram_width
+
+struct extract {
+    std::vector<std::string> upper_bounds_;
+    std::vector<std::string> lower_bounds_;
+    std::vector<int> values_;
+    unsigned int size() const {return values_.size();}
+};
+
+struct visualization_data {
+  std::vector<std::string> str_values_;
+  std::vector<int> scale_factors_;
+
+  size_t lower_bounds_width_ = 0; 
+  size_t upper_bounds_width_ = 0;
+  size_t str_values_width_ = 0;
+  size_t external_line_shift_ = 0; 
+  visualization_data(const std::vector<std::string>& str_values, 
+                     const std::vector<int>& scale_factors,
+                     const size_t& lower_bounds_width,
+                     const size_t& upper_bounds_width, 
+                     const size_t& str_values_width,
+                     const size_t& external_line_shift)
+                     :
+                     str_values_{str_values}, 
+                     scale_factors_{scale_factors},
+                     lower_bounds_width_{lower_bounds_width},
+                     upper_bounds_width_{upper_bounds_width},
+                     str_values_width_{str_values_width},
+                     external_line_shift_{external_line_shift}
+                     {}
+};
+
+template <class histogram>
+extract extract_data(const histogram& h) {
+  std::string lower, upper;
+
+  auto data = indexed(h, coverage::all);
+
+  extract ex;  
+  for (auto x : data) {
+    lower = str( boost::format("%.1f") % x.bin().lower() );
+    upper = str( boost::format("%.1f") % x.bin().upper() );
+    ex.lower_bounds_.push_back(lower);
```

Comment:
where is `str` coming from? you will need to rewrite this. doing one pass over the histogram to figure out all the widths and then a second pass where you actually generate all the text strings.

The main function should not generate a string, because a string has to be allocated. The idiomatic way is to accept a `std::ostream&` and return it. Look how it is done in the `ostream.hpp` headers. The main function should have this signature `template <class Histogram> std::ostream& display(std::ostream& os, const Histogram& h, unsigned terminal_width=0)`. The value `0` indicates that the library should an internal default value. If the user passes a non-zero value, then this should be used as the terminal width. For now, you can use the 60 character limit as the default value. I will implement some terminal width detection code later to change that to something adapted to the current terminal.